sdk/src/ed25519_instruction.rs

```diff
@@ -0,0 +1,287 @@
+#![cfg(feature = "full")]
+
+use crate::instruction::Instruction;
+use ed25519_dalek::{ed25519::signature::Signature, Signer, Verifier};
+use serde_derive::{Deserialize, Serialize};
+
+#[derive(Debug, PartialEq)]
+pub enum Ed25519Error {
+    InvalidPublicKey,
+    InvalidSignature,
+    InvalidDataOffsets,
+    InvalidInstructionDataSize,
+}
+
+pub const PUBKEY_SERIALIZED_SIZE: usize = 32;
+pub const SIGNATURE_SERIALIZED_SIZE: usize = 64;
+pub const SIGNATURE_OFFSETS_SERIALIZED_SIZE: usize = 11;
+pub const DATA_START: usize = SIGNATURE_OFFSETS_SERIALIZED_SIZE + 1;
+
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct Ed25519SignatureOffsets {
+    signature_offset: u16,            // offset to ed25519 signature of 64 bytes
+    signature_instruction_index: u8,  // instruction index to find signature
+    public_key_offset: u16,           // offset to public key of 32 bytes
+    public_key_instruction_index: u8, // instruction index to find public key
+    message_data_offset: u16,         // offset to start of message data
+    message_data_size: u16,           // size of message data
+    message_instruction_index: u8,    // index of instruction data to get message data
+}
+
+pub fn new_ed25519_instruction(keypair: &ed25519_dalek::Keypair, message: &[u8]) -> Instruction {
+    let signature = keypair.sign(message).to_bytes();
+    let pubkey = keypair.public.to_bytes();
+
+    assert_eq!(pubkey.len(), PUBKEY_SERIALIZED_SIZE);
+    assert_eq!(signature.len(), SIGNATURE_SERIALIZED_SIZE);
+
+    let mut instruction_data = vec![];
+    instruction_data.resize(
+        DATA_START
+            .saturating_add(signature.len())
+            .saturating_add(pubkey.len())
+            .saturating_add(message.len())
+            .saturating_add(1),
+        0,
+    );
+
+    let public_key_offset = DATA_START;
+    instruction_data[public_key_offset..public_key_offset.saturating_add(pubkey.len())]
+        .copy_from_slice(&pubkey);
+
+    let signature_offset = DATA_START.saturating_add(pubkey.len());
+    instruction_data[signature_offset..signature_offset.saturating_add(signature.len())]
+        .copy_from_slice(&signature);
+
+    let message_data_offset = signature_offset
+        .saturating_add(signature.len())
+        .saturating_add(1);
+    instruction_data[message_data_offset..].copy_from_slice(message);
```

Comment:
These are guarenteed to be the same length?

Reply:
Well, they should be. I've added `message_data_offset + message.len()` to be sure